### PR TITLE
Fix issues when active and passive changes in HA

### PR DESCRIPTION
**Pull request by ramindu90**
## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/ha/HACoordinationSourceHandler.java [hunk 1]
```diff
@@ -154,85 +154,95 @@ public String getElementId() {
...
+            if (eventSyncConnection != null) {
+                QueuedEvent queuedEvent = new QueuedEvent(siddhiAppName, sourceHandlerElementId, sequenceIDGenerator
+                        .incrementAndGet(), event, transportSyncProperties);
                 try {
-                    eventSyncConnection.send(HAConstants.CHANNEL_ID_MESSAGE, messageBuffer.array());
-                    if (throughputTracker != null && StreamProcessorDataHolder.isStatisticsEnabled()) {
-                        throughputTracker.eventIn();
+                    messageBuffer = BinaryEventConverter.convertToBinaryMessage(new QueuedEvent[]{queuedEvent});
+                } catch (IOException e) {
+                    log.error("Error in converting events to binary message.Hence not sending message to the passive node");
```
Comment (mohanvive) [inline components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/ha/HACoordinationSourceHandler.java]: Any reason on, why can't we add the error trace here? This is a genuine error which we should print the trace as well. Isn't it?

File: components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/ha/HACoordinationSourceHandler.java [hunk 2]
```diff
@@ -154,85 +154,95 @@ public String getElementId() {
...
+                    log.error("Error in returning the tcpClient connection object to the pool. ", e);
                 }
             }
-            try {
-                eventSyncConnectionPoolFactory.returnObject(HAConstants.ACTIVE_NODE_CONNECTION_POOL_ID, eventSyncConnection);
-            } catch (Exception e) {
-                log.error("Error in returning the tcpClient connection object to the pool. ", e);
-            }
+        } catch (ConnectionUnavailableException e) {
+            log.warn(e.getMessage());
```
Comment (mohanvive) [inline components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/ha/HACoordinationSourceHandler.java]: Let's provide a bit more context and append the error message with that.

File: components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/ha/HACoordinationSourceHandler.java [hunk 3]
```diff
@@ -154,85 +154,95 @@ public String getElementId() {
...
-                log.error("Error in converting events to binary message.Hence not sending message to the passive node");
-            }
-            if (messageBuffer != null) {
                 try {
-                    eventSyncConnection.send(HAConstants.CHANNEL_ID_MESSAGE, messageBuffer.array());
-                    if (throughputTracker != null && StreamProcessorDataHolder.isStatisticsEnabled()) {
-                        throughputTracker.eventsIn(events.length);
+                    messageBuffer = BinaryEventConverter.convertToBinaryMessage(queuedEvents);
+                } catch (IOException e) {
+                    log.error("Error in converting events to binary message.Hence not sending message to the passive node");
```
Comment (mohanvive) [inline components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/ha/HACoordinationSourceHandler.java]: error trace in the log?

File: components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/ha/HACoordinationSourceHandler.java [hunk 4]
```diff
@@ -154,85 +154,95 @@ public String getElementId() {
...
+                    log.error("Error in returning the tcpClient connection object to the pool. ", e);
                 }
             }
-            try {
-                eventSyncConnectionPoolFactory.returnObject(HAConstants.ACTIVE_NODE_CONNECTION_POOL_ID, eventSyncConnection);
-            } catch (Exception e) {
-                log.error("Error in returning the tcpClient connection object to the pool. ", e);
-            }
+        } catch (ConnectionUnavailableException e) {
+            log.warn(e.getMessage());
```
Comment (mohanvive) [inline components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/ha/HACoordinationSourceHandler.java]: add more context to error.

File: components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/persistence/PersistenceManager.java [hunk 5]
```diff
@@ -88,52 +88,60 @@ private void persist() {
...
+                try {
+                    if (fullStateFuture != null) {
+                        fullStateFuture.get(60000, TimeUnit.MILLISECONDS);
+                    } else {
+                        for (Future future : persistenceReference.getIncrementalStateFuture()) {
+                            future.get(60000, TimeUnit.MILLISECONDS);
+                        }
                     }
+                } catch (Throwable e) {
+                    log.error("Active Node: Persisting of Siddhi app is not successful. Check if app deployed properly");
```
Comment (mohanvive) [inline components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/persistence/PersistenceManager.java]: Don't we need error trace here?

File: components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/ha/transport/EventSyncConnection.java [hunk 6]
```diff
@@ -67,63 +63,72 @@ protected void initChannel(SocketChannel ch) throws Exception {
...
+        public void shutdown() {
+            if (channel != null && channel.isOpen()) {
+                try {
+                    channel.close();
+                    channel.closeFuture().sync();
+                } catch (InterruptedException e) {
+                    log.error("Error closing connection to '" + hostAndPort + "' from client '" + sessionId +
+                            "', " + e);
+                }
+                channel.disconnect();
```
Comment (suhothayan) [inline components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/ha/transport/EventSyncConnection.java]: This should be in the finally block

File: components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/ha/HACoordinationSourceHandler.java (hunk 1 above)
Comment (ramindu90) [reply components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/ha/HACoordinationSourceHandler.java]: added the trace

File: components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/ha/HACoordinationSourceHandler.java (hunk 4 above)
Comment (ramindu90) [reply components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/ha/HACoordinationSourceHandler.java]: added.

File: components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/persistence/PersistenceManager.java (hunk 5 above)
Comment (ramindu90) [reply components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/persistence/PersistenceManager.java]: added.

File: components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/ha/transport/EventSyncConnection.java (hunk 6 above)
Comment (ramindu90) [reply components/org.wso2.carbon.stream.processor.core/src/main/java/org/wso2/carbon/stream/processor/core/ha/transport/EventSyncConnection.java]: added the finally block

**Pull request merged**